File: Downloads/SystemCoupling/syc_ParticipantLibrary_r2_24/PipeMapping/PipeMapping.py
'''
Copyright ANSYS, Inc. Unauthorized use, distribution, or duplication is prohibited.
'''

'''
This is the pipe mapping script.

It generates two different meshes, each for a region representing a 
pipe geometry. It then performs non-conformal mesh mapping between 
these two regions.

The mesh refinement level, pipe offset, pipe radius, etc. can be 
controlled by the command line arguments. The resulting arrays can 
be written for further post-processing.
'''
import os
import sys

def addSyCDLLs():
    if sys.platform.startswith("win"):
        for p in os.environ["PYTHON_DLL_PATH"].split(os.pathsep):
            try:
                os.add_dll_directory(p)
            except (FileNotFoundError, OSError):
                pass  # skip any paths that don't exist
                
addSyCDLLs()

from PipeMeshGenerator import getMeshProperties
from PipeMeshGenerator import getNodeCoords
from PipeMeshGenerator import getElemNodeCounts
from PipeMeshGenerator import getElemNodeIds
from PipeMeshGenerator import getSolutionData
from PipeMeshGenerator import initializeMesh

from pyExt import SystemCouplingParticipant as sysc
import time

def getSurfaceMesh(regionName):
    ''' Function that returns surface mesh, given a region. '''
    ncs = sysc.OutputVectorData(getNodeCoords(regionName))
    encs= sysc.OutputIntegerData(getElemNodeCounts(regionName))
    enis= sysc.OutputIntegerData(getElemNodeIds(regionName))
    nodes = sysc.NodeData(ncs)
    faces = sysc.FaceData(
        sysc.ElementNodeCountData(encs),
        sysc.ElementNodeConnectivityData(enis))
    return sysc.SurfaceMesh(nodes, faces)

def getInputScalarData(regionName, variableName):
    '''
    Function that returns scalar input variable values,
    given a region and a variable.
    '''
    return sysc.InputScalarData(getSolutionData(regionName, variableName))

def getOutputScalarData(regionName, variableName):
    '''
    Function that returns scalar output variable values,
    given a region and a variable.
    '''
    return sysc.OutputScalarData(getSolutionData(regionName, variableName))

def printMaxError(regionName):
    ''' Function to calculate and print max error on a given region. '''
    expectedVar = "linear1" if regionName == "quad" else "linear2"
    testVar = "linear1" if expectedVar == "linear2" else "linear2"
    expectedValues = getSolutionData(regionName, expectedVar)
    testValues = getSolutionData(regionName, testVar)
    numVals = len(testValues)
    maxError = 0.0
    for n in range(numVals):
        testValue = testValues[n]
        expectedValue = expectedValues[n]
        maxError = max(maxError, abs(expectedValue - testValue))
    print(f"Maximum error on {regionName}: {maxError:.4e}")

'''
Parse the command line arguments.
'''
import argparse
parser = argparse.ArgumentParser()
parser.add_argument('--quad', type=int, default=20, help='Refinement level for quad region.')
parser.add_argument('--tri', type=int, default=30, help='Refinement level for tri region.')
parser.add_argument('--offset', type=float, default=0.0, help='Axial offset for quad region.')
parser.add_argument('--qradius', type=float, default=0.05, help='Pipe radius for quad region.')
parser.add_argument('--tradius', type=float, default=0.05, help='Pipe radius for tri region.')
parser.add_argument('--write', type=int, default=1, help='Flag to turn on/off .dat file writing.')
'''
Add overlap layer for consistency 
# (it has no effect here, since this script can only run in serial).
'''
parser.add_argument('--overlap', type=int, default=0, help='Overlap layer (not applicable).')
args = parser.parse_known_args()[0]

meshProperties = getMeshProperties()
meshProperties.quadRefine = args.quad
meshProperties.triRefine = args.tri
meshProperties.pipeOffset = args.offset
meshProperties.quadPipeRadius = args.qradius
meshProperties.triPipeRadius = args.tradius

print('Setup:')
print(f'Quad refinement: {meshProperties.quadRefine}')
print(f'Tri refinement: {meshProperties.triRefine}')
print(f'Pipe offset: {meshProperties.pipeOffset}')
print(f'Quad radius: {meshProperties.quadPipeRadius}')
print(f'Tri radius: {meshProperties.triPipeRadius}')
print(f'Write: {args.write}')
print("==========")

'''
Create SystemCoupling object.
'''
sc = sysc.SystemCoupling()

''' Add variables, regions, and the interface. '''
variableLinear1 = sysc.Variable("linear1", sysc.Scalar, False, sysc.Node)
variableLinear2 = sysc.Variable("linear2", sysc.Scalar, False, sysc.Node)

quadRegion = sysc.Region("quad", sysc.Surface)
quadRegion.addOutputVariable(variableLinear1)
quadRegion.addInputVariable(variableLinear2)

triRegion = sysc.Region("tri", sysc.Surface)
triRegion.addOutputVariable(variableLinear2)
triRegion.addInputVariable(variableLinear1)

couplingInterface = sysc.CouplingInterface("interface")
couplingInterface.addSideOneRegion(quadRegion)
couplingInterface.addSideTwoRegion(triRegion)

sc.addCouplingInterface(couplingInterface)

''' Register data access functions. '''
sc.registerSurfaceMeshAccess(getSurfaceMesh)
sc.registerInputScalarDataAccess(getInputScalarData)
sc.registerOutputScalarDataAccess(getOutputScalarData)

''' Initialize mesh and solution data on both sides of the interface. '''
startTime = time.time()
initializeMesh()
print(f"Time to generate mesh: {time.time() - startTime:.4e} [s]")

''' Update inputs - this will perform mapping and bring inputs up-to-date. '''
startTime = time.time()
sc.updateInputs()
print(f"Time to update inputs: {time.time() - startTime:.4e} [s]")

''' Write results, if desired. '''
if bool(args.write):
    startTime = time.time()
    sc.writeResults(sysc.ResultsInfo("pipe"))
    print(f"Time to write results: {time.time() - startTime:.4e} [s]")

printMaxError("quad")
printMaxError("tri")

File: Downloads/SystemCoupling/syc_ParticipantLibrary_r2_24/PipeMapping/src_cpp/PipeMapping.cpp
/*
 * Copyright ANSYS, Inc. Unauthorized use, distribution, or duplication is prohibited.
 */

#include "PipeMeshGenerator.hpp"

#include "SystemCouplingParticipant/SystemCoupling.hpp"

#include <mpi.h>

#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

// Define parallel properties.
int myrank = 0;
int numranks = 1;

/* Function that returns surface mesh, given a region. */
sysc::SurfaceMesh getSurfaceMesh(
  const std::string& regionName)
{
  return sysc::SurfaceMesh(
    sysc::NodeData(sysc::OutputVectorData(getNodeCoords(regionName))),
    sysc::FaceData(
      sysc::ElementNodeCountData(sysc::OutputScalarData(getElemNodeCounts(regionName))),
      sysc::ElementNodeConnectivityData(sysc::OutputScalarData(getElemNodeIds(regionName)))));
}

/*    
 * Function that returns scalar input variable values,
 * given a region and a variable.
 */
sysc::InputScalarData getInputScalarData(
  const std::string& regionName,
  const std::string& variableName)
{
  return sysc::InputScalarData(getSolutionData(regionName, variableName));
}

/*
 * Function that returns scalar output variable values,
 * given a region and a variable.
 */
sysc::OutputScalarData getOutputScalarData(
  const std::string& regionName,
  const std::string& variableName)
{
  return sysc::OutputScalarData(getSolutionData(regionName, variableName));
}

/* Function to calculate and print max error on a given region. */
void printMaxError(const std::string& regionName)
{
  std::string expectedVar = regionName == "quad" ? "linear1" : "linear2";
  std::string testVar = expectedVar == "linear2" ? "linear1" : "linear2";
  const std::vector<double>& expectedValues = getSolutionData(regionName, expectedVar);
  const std::vector<double>& testValues = getSolutionData(regionName, testVar);
  std::size_t numVals = testValues.size();
  double maxError = 0.0, totalMaxError;
  for (std::size_t n = 0; n < numVals; ++n) {
    const double& testValue = testValues[n];
    const double& expectedValue = expectedValues[n];
    maxError = std::fmax(maxError, std::abs(expectedValue - testValue));
  }

  MPI_Reduce(&maxError, &totalMaxError, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);

  if (myrank == 0) {
    std::cout << "Maximum error on " << regionName << ": " << totalMaxError << "\n";
  }
}

/* Function to write out mesh and solution data arrays to text files. */
void writeArrays(const std::string& regionName)
{
  const std::vector<double>& currNodeCoords = getNodeCoords(regionName);
  const std::vector<int>& currElemNodeCounts = getElemNodeCounts(regionName);
  const std::vector<int>& currElemNodeIds = getElemNodeIds(regionName);

  std::string fileName = regionName + "." + std::to_string(myrank) + ".dat";
  std::ofstream fileHandle(fileName.c_str());
  fileHandle << std::setprecision(8) << std::scientific;

  // Write number of nodes, elements, and solution data arrays.
  int numNodes = static_cast<int>(currNodeCoords.size() / 3);
  int numElems = static_cast<int>(currElemNodeCounts.size());
  int numScalarNodeArrays = 2;  // "linear1" and "linear2"
  fileHandle << numNodes << ", "
             << numElems << ", "
             << numScalarNodeArrays << "\n";

  // Write node coordinates.
  for (int n = 0; n < numNodes; ++n) {
    fileHandle << currNodeCoords[n * 3 + 0] << ", ";
    fileHandle << currNodeCoords[n * 3 + 1] << ", ";
    fileHandle << currNodeCoords[n * 3 + 2] << "\n";
  }

  // Write element-to-node connectivity.
  std::size_t elemNodeIndex = 0;
  for (int e = 0; e < numElems; ++e) {
    // Write the first node id for the current element.
    fileHandle << currElemNodeIds[elemNodeIndex];
    ++elemNodeIndex;
    // Write subsequent node ids for the current element.
    for (int en = 1; en < currElemNodeCounts[e]; ++en) {
      fileHandle << ", " << currElemNodeIds[elemNodeIndex];
      ++elemNodeIndex;
    }
    fileHandle << "\n";
  }

  // Write solution data arrays.
  std::vector<std::string> variableNames(2);
  variableNames[0] = "linear1";
  variableNames[1] = "linear2";
  for (const std::string& variableName : variableNames) {
    const std::vector<double>& currData = getSolutionData(regionName, variableName);
    fileHandle << variableName << "\n";
    for (int n = 0; n < numNodes; ++n) {
      fileHandle << currData[n] << "\n";
    }
  }
}

/*
 * This is the main function for the pipe mapping program.
 *
 * It generates two different meshes, each for a region representing a pipe
 * geometry. It then performs non-conformal mesh mapping between these two
 * regions.
 *
 * The mesh refinement level, pipe offset, pipe radius, etc. can be controlled
 * by the command line arguments. The resulting arrays can be written for
 * further post-processing.
 *
 */
int main(int argc, char* argv[])
{
  try {
    // Parse command line arguments.
    bool write = true;

    PipeMeshProperties& meshProperties = getMeshProperties();

    for (int i = 1; i < argc; ++i) {
      if (std::string(argv[i]) == "--quad" && i + 1 < argc) {
        meshProperties.quadRefine = std::stoi(argv[i + 1]);
      }
      else if (std::string(argv[i]) == "--tri" && i + 1 < argc) {
        meshProperties.triRefine = std::stoi(argv[i + 1]);
      }
      else if (std::string(argv[i]) == "--offset" && i + 1 < argc) {
        meshProperties.pipeOffset = std::stod(argv[i + 1]);
      }
      else if (std::string(argv[i]) == "--qradius" && i + 1 < argc) {
        meshProperties.quadPipeRadius = std::stod(argv[i + 1]);
      }
      else if (std::string(argv[i]) == "--tradius" && i + 1 < argc) {
        meshProperties.triPipeRadius = std::stod(argv[i + 1]);
      }
      else if (std::string(argv[i]) == "--write" && i + 1 < argc) {
        write = static_cast<bool>(std::stoi(argv[i + 1]));
      }
      else if (std::string(argv[i]) == "--overlap" && i + 1 < argc) {
        meshProperties.overlapLayers = std::stoi(argv[i + 1]);
      }
    }

    // Initialize MPI.
    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
    MPI_Comm_size(MPI_COMM_WORLD, &numranks);

    meshProperties.myrank = myrank;
    meshProperties.numranks = numranks;

    if (myrank == 0) {
      std::cout << "Setup:\n";
      std::cout << "Quad refinement: " << meshProperties.quadRefine << '\n';
      std::cout << "Tri refinement: " << meshProperties.triRefine << '\n';
      std::cout << "Pipe offset: " << meshProperties.pipeOffset << '\n';
      std::cout << "Quad radius: " << meshProperties.quadPipeRadius << '\n';
      std::cout << "Tri radius: " << meshProperties.triPipeRadius << '\n';
      std::cout << "Write: " << write << '\n';
      std::cout << "Overlap layers: " << meshProperties.overlapLayers << '\n';
      std::cout << "Processes: " << numranks << '\n';
      std::cout << "==========\n";
    }

    // Create SystemCoupling object.
    sysc::SystemCoupling sc;

    // Add variables, regions, and the interface.
    sysc::Variable variableLinear1("linear1", sysc::Scalar, false, sysc::Node);
    sysc::Variable variableLinear2("linear2", sysc::Scalar, false, sysc::Node);

    sysc::Region quadRegion("quad", sysc::Surface);
    quadRegion.addOutputVariable(variableLinear1);
    quadRegion.addInputVariable(variableLinear2);

    sysc::Region triRegion("tri", sysc::Surface);
    triRegion.addOutputVariable(variableLinear2);
    triRegion.addInputVariable(variableLinear1);

    sysc::CouplingInterface couplingInterface("interface");
    couplingInterface.addSideOneRegion(quadRegion);
    couplingInterface.addSideTwoRegion(triRegion);

    sc.addCouplingInterface(couplingInterface);

    // Register data access functions.
    sc.registerSurfaceMeshAccess(&getSurfaceMesh);
    sc.registerInputScalarDataAccess(&getInputScalarData);
    sc.registerOutputScalarDataAccess(&getOutputScalarData);

    double startTime, localTime, totalTime;
    std::cout << std::setprecision(4) << std::scientific;

    // Initialize mesh and solution data on both sides of the interface.
    startTime = MPI_Wtime();
    initializeMesh();
    localTime = MPI_Wtime() - startTime;
    MPI_Reduce(&localTime, &totalTime, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
    if (myrank == 0) {
      std::cout << "Time to generate mesh: " << totalTime << " [s]\n";
    }

    // Update inputs - this will perform mapping and bring inputs up-to-date.
    startTime = MPI_Wtime();
    sc.updateInputs();
    localTime = MPI_Wtime() - startTime;
    MPI_Reduce(&localTime, &totalTime, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
    if (myrank == 0) {
      std::cout << "Time to update inputs: " << totalTime << " [s]\n";
    }

    // Write results, if desired.
    if (write) {
      startTime = MPI_Wtime();
      writeArrays("quad");
      writeArrays("tri");
      sc.writeResults(sysc::ResultsInfo("pipe"));
      localTime = MPI_Wtime() - startTime;
      MPI_Reduce(&localTime, &totalTime, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);
      if (myrank == 0) {
        std::cout << "Time to write results: " << totalTime << " [s]\n";
      }
    }

    printMaxError("quad");
    printMaxError("tri");
  }
  catch (const std::exception& exc) {
    std::cout << "ERROR: " << exc.what() << std::endl;
    MPI_Finalize();
    return EXIT_FAILURE;
  }

  MPI_Finalize();
  return EXIT_SUCCESS;
}
